### PR TITLE
Pilip/identity

### PR DESCRIPTION
**Pull request by blastrock**

### CODE REVIEW

File: identity/generate.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+type PublicIdentity struct {
+	TrustchainID []byte `json:"trustchain_id"`
+	Target       string `json:"target"`
+	Value        []byte `json:"value"`
+}
+
+//Create a user token for given user.
+func Create(config Config, userID string) (string, error) {
```
Comment (maximerety) [inline identity/generate.go]: `Create` or keep `Generate` as before and as described in the README?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      return "", error.New("Unauthorized")
+    }
+
+    // Retrieve a previously stored identity for this user
+    tkIdentity := retrieveIdentity(userID)
+    if tkIdentity == "" {
+      return "", error.New("Not found")
+    }
+
+    publicIdentity, err := identity.getPublicIdentity(tkIdentity)
```
Comment (maximerety) [inline README.md]: GetPublicIdentity with a big G

File: identity/generate.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+	b64PublicIdentity, err := toB64JSON(publicIdentity)
+	if err != nil {
+		return "", err
+	}
+
+	return b64PublicIdentity, nil
+}
+
+func generateToken(trustchainID []byte, trustchainPrivateKey []byte,
```
Comment (maximerety) [inline identity/generate.go]: Isn't `Token` a confusing naming?

File: identity/generate.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+		Value:                        userID,
+		UserSecret:                   userSecret,
+		TrustchainID:                 trustchainID,
+	}
+
+	b64Token, err := toB64JSON(identity)
+	if err != nil {
+		return "", err
+	}
+	return b64Token, nil
```
Comment (maximerety) [inline identity/generate.go]: Or return `toB64JSON(identity)` directly? No need to check output!

File: identity/generate.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+		Target:       "user",
+		Value:        identity.Value,
+	}
+
+	b64PublicIdentity, err := toB64JSON(publicIdentity)
+	if err != nil {
+		return "", err
+	}
+
+	return b64PublicIdentity, nil
```
Comment (maximerety) [inline identity/generate.go]: Or return `toB64JSON(publicIdentity)` directly? No need to check output!

File: identity_test/generate_test.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		err = json.Unmarshal(sidentity, &identity)
+		Expect(err).To(Not(HaveOccurred()))
+
+		Expect(identity.TrustchainID).To(Equal(trustchainID))
+		Expect(identity.Target).To(Equal("user"))
+		Expect(identity.Value).To(Equal(obfuscatedUserID))
+		checkDelegationSignature(identity, trustchainPublicKey)
+	})
+
+	It("returns a tanker public identity from an tanker indentity", func() {
```
Comment (maximerety) [inline identity_test/generate_test.go]: a**n** tanker i**n**dentity -> a tanker identity (too many n's!)

File: README.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        // Store the newly generated identity
+        storeIdentity(userID, identity)
+    }
+
+    // From now, the same identity will always be returned to a given user
+    return identity, nil
+}
+```
+
+Read more about identities in the [Tanker guide](https://tanker.io/docs/latest/guide/server/).
```
[the remark this answers is not in the document]
Comment (blastrock) [reply README.md]: linked to user-token for the time being

[the remark this answers is not in the document]
Comment (maximerety) [reply README.md]: I get a 404

**Pull request merged**